nexus/src/app/background/tasks/instance_reincarnation.rs

```diff
@@ -0,0 +1,715 @@
+                    "already_reincarnated" => status.already_reincarnated.len(),
+                );
+            } else {
+                debug!(
+                    &opctx.log,
+                    "instance reincarnation completed; no instances \
+                     in need of reincarnation";
+                    "instances_found" => status.instances_found,
+                    "already_reincarnated" => status.already_reincarnated.len(),
+                );
+            };
+            serde_json::json!(status)
+        })
+    }
+}
+
+impl InstanceReincarnation {
+    pub(crate) fn new(
+        datastore: Arc<DataStore>,
+        sagas: Arc<dyn StartSaga>,
+        default_cooldown: Duration,
+    ) -> Self {
+        let default_cooldown = TimeDelta::from_std(default_cooldown)
+            .expect("duration should be in range");
+        Self { datastore, sagas, default_cooldown }
+    }
+
+    async fn actually_activate(
+        &mut self,
+        opctx: &OpContext,
+        status: &mut InstanceReincarnationStatus,
+    ) {
+        let mut tasks = JoinSet::new();
+        let mut paginator =
+            Paginator::new(nexus_db_queries::db::datastore::SQL_BATCH_SIZE);
+
+        while let Some(p) = paginator.next() {
+            let maybe_batch = self
+                .datastore
+                .find_reincarnatable_instances(opctx, &p.current_pagparams())
+                .await;
+            let batch = match maybe_batch {
+                Ok(batch) => batch,
+                Err(error) => {
+                    error!(
+                        opctx.log,
+                        "failed to list instances in need of reincarnation";
+                        "error" => &error,
+                    );
+                    status.query_error = Some(error.to_string());
+                    break;
+                }
+            };
+
+            paginator = p.found_batch(&batch, &|instance| instance.id());
+
+            let found = batch.len();
+            if found == 0 {
+                debug!(
+                    opctx.log,
+                    "no more instances in need of reincarnation";
+                    "total_found" => status.instances_found,
+                );
+                break;
+            }
+
+            let prev_sagas_started = tasks.len();
+            status.instances_found += found;
+
+            let serialized_authn = authn::saga::Serialized::for_opctx(opctx);
+            for db_instance in batch {
+                let instance_id = db_instance.id();
+
+                // Don't reincarnate instances that were restarted too recently.
+                //
+                // Note that we *could* filter out instances that reincarnated
+                // too recently in the `find_reincarnatable_instances` database
+                // query, but I thought it was nice to be able to include them
+                // in this background task's OMDB status.
```

Comment:
Feel free to ignore this or defer it: my inclination here would be to filter it out in the db query and provide a separate omdb command that looks at the database state and reports instances in the cooldown period.  The reason is mostly that when I've had processes like this that use queries where not all the rows are actionable, I've run into trouble when the non-actionable rows end up dominating the result set.  But that does seem pretty unlikely here.

Reply:
I think you're right that filtering them out in the query is definitely nicer, and if we start using the size of the result set as a rudimentary concurrency limit...well, ineligible instances should probably not count against that. I might change it to do that, and defer adding an OMDB command specifically for reincarnation states for a subsequent PR? Or maybe I'll just go do it now...

Reply:
I changed the query to filter out instances that are still in the cooldown period in https://github.com/oxidecomputer/omicron/pull/6503/commits/5282d575aab5245add57a45c6aca2c64a116db46.